File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  const newArr =  [...Array(arr.length).keys()]
  const duplicate = arr.find((elem, index) => {
    index !== newArr[index]
    return elem
  })
  if(!duplicate){
    return -1
  }
  return duplicate
}

if (require.main === module) {
  let arr1 = ['a','b','c','d','e','c','f','g','a']
  console.log("Expecting: c");
  console.log("=>", findFirstDuplicate(arr1));

  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));
  
  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

//pseudocode:
// function findFirstDuplicate
//   create Array populated up to the length of arr
//   compare the new array value with the index of arr
//   return value if not equal
//   if no value returned return -1
// done


// And a written explanation of your solution
//
// I create a new array with a range of 0..arr.length, so now I have the argument 
// array and a new array, I then iterate through the arr with .find(),
// I compare the index of each element to the value of the new array 
// knowing the arr.indexOf() function will return the index of the first 
// instance of the element in the array. If there a duplicate the index and the 
// new array value will not be equal returning the element.
//

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  arr2 = []
  duplicate = arr.each do |x|
    if  arr2.include?(x)
      return x;
    end
    arr2<<x
  end
 -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!

  puts "Expecting: 7"
  puts "=>", find_first_duplicate([1, 7, 2, 3, 4, 7, 4])
  
  puts

  puts "Expecting: 9"
  puts "=>", find_first_duplicate([1, 9, 2, 5, 7, 8, 9, 6])
  
  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([7, 4, 6, 9])
  
  puts
end

# Please add your pseudocode to this file
# 
# find_first_duplicate(arr)
#   create new empty array 
#   iterate over arr 
#   if new array includes element 
#     return element
#   if not add element to new array
#   if no duplicates are found return -1
# done


# And a written explanation of your solution
# 
# I begin by creating an empty array, then I iterate over argument array
# if the element the iterator is working on is included in the new array
# then I want the method to return the element. If the element isnt included
# in the new array then push that element to the new array and continue iterating
# over the rest of the elements until either a duplicate is found or if not return 
# the value of -1
